Add Langfile Duplicate Identifier script for GitHook

pbechtold opened the pull request:
Hey everyone,

This PR adds the Bash script check_langfile_duplicate_identifier.sh to be included in the GitHooks. The PR description also includes the necessary lines for /captainhook.json and .github/workflows/checks.yml.

Best regards!

1. To activate, these lines must be in the "/captainhook.json" file, in line 66
 ```
        {
        "action": "scripts/check_langfile_duplicate_identifier.sh",
        "options": [],
        "conditions": [
          {
            "exec": "\\CaptainHook\\App\\Hook\\Condition\\FileStaged\\Any",
            "args": [
              [
                "*.lang"
              ]
            ]
          }
        ]
      },
```
	  

2. To activate, these lines must be in the ".github/workflows/checks.yml" file - line 58  
```
      - name: Language Duplicate Identifier Check
        run: CI/check_langfile_duplicate_identifier.sh
        env:
          GHRUN: "yes"
          PR_NUMBER: ${{ github.event.number }}
          GH_SHA: ${{ github.sha }}
```

pbechtold committed:
Add check_langfile_duplicate_identifier script

pbechtold committed:
Update check_langfile_duplicate_identifier.sh

matthiaskunkel commented:
PR is related to Mantis41710: Guideline violations: duplicate variable_IDs in various components. Once the 291 violations have been removed, this PR will help to get rid of duplicate variable_IDs in our language file.

My suggestion for the next steps would be:
1. Solving all noticed violations (and closing Mantis bug). This would be done by the language maintainer. I would rename the IDs and adapt the strings in the GUI classes.
2. Once all violations are fixed, I would merge the PR to activate the GitHook. After this moment, it will no longer be possible to commit a language file with duplicate variable_IDs.

matthiaskunkel commented:
Cleaning of lang files and renaming variables will only be made for trunk.

iszmais commented:
@matthiaskunkel 

is this still looking for shepherd?
I'm asking since you are assigned.

Greetings,
@iszmais

matthiaskunkel commented:
Yes, we still look for a shephard who checks the scripts. No need to merge them. I'll gonna do that once I have removed the ~300 violations that exist. But a 'thumb-up' from you – or change requests – would be highly appreciated.

iszmais commented:
@matthiaskunkel Thanks for your response.

Unfortunately, I would not consider myself an expert in sh script and therefore qualified for a review. But i will transmit you request to so potential candidates.

Greetings,
@iszmais

lscharmer commented on scripts/check_langfile_duplicate_identifier.sh:
```diff
@@ -0,0 +1,36 @@
+  # Find the duplicate identifiers
+  duplicate_identifiers=$(sed -n '/<!-- language file start -->/,$p' $file | awk -F '#:#' '{print tolower($2)}' | sort -f | uniq -d)
+
+  if [[ -n $duplicate_identifiers ]]; then
```
Please quote the variable `$duplicate_identifiers` to prevent argument expansion:
```bash
duplicate_identifiers='-f foo'
if [[ -n $duplicate_identifiers ]]; then echo ok; fi
# bash: [: -f: binary operator expected
```

And as a suggestion: I would remove the extra pair of square braces (the `-n` flag is the same with `[` but no bashism).

```suggestion
  if [ -n "$duplicate_identifiers" ]; then
```

lscharmer commented on scripts/check_langfile_duplicate_identifier.sh:
```diff
@@ -0,0 +1,36 @@
+     # Loop through each duplicate identifier
+     for duplicate_identifier in $duplicate_identifiers; do
+        # Find all lines containing this exact duplicate and print the modules
+        IFS=$'\n' lines_with_duplicate=($(sed -n '/<!-- language file start -->/,$p' $file | awk -F '#:#' -v id="$duplicate_identifier" 'tolower($2) == tolower(id){print $1}'))
```
Please quote `$file` to prevent argument expansion:
The use of bash arrays can be removed, as it has no effect because spaces and newlines are treated the same anyway in line `27`

```suggestion
        lines_with_duplicate=$(sed -n '/<!-- language file start -->/,$p' "$file" | awk -F '#:#' -v id="$duplicate_identifier" 'tolower($2) == tolower(id){print $1}')

```
and
`${lines_with_duplicate[@]} -> ${lines_with_duplicate}`
in line `27`

iszmais commented:
Thank you @lscharmer for offering your help to a qualified review here!

lscharmer commented on scripts/check_langfile_duplicate_identifier.sh:
```diff
@@ -0,0 +1,36 @@
+  fi
+done
+
+exit $exit_status
```
This is just a suggestion:
Can be removed you can just directly `exit` in line `32` the default exit code is `0`.

lscharmer commented on scripts/check_langfile_duplicate_identifier.sh:
```diff
@@ -0,0 +1,36 @@
+     done
+
+     # Mark that duplicates are found
+     exit_status=127
```
Why exit code `127`? Generally this is ok but `127` is used by `bash` when a command could not be found:
```bash
$ hejhooo
hejhoo: command not found
$ echo $?
127
```

So I would suggest to just exit with `1` instead:

```suggestion
     exit 1
```

lscharmer reviewed (changes requested):
Hi @pbechtold,
thank you for this PR :)

This script looks generally good to me, there are only a few things that should be changed. Additionally all comments marked as a suggestion don't need to be implemented but would be an improvement IMO.

Please implement the following:
- [x] Please use english for console output :)

And as a suggestion:
To harden this script I would use `echo "$foo" | while read bar; do` instead of `for foo in $bar; do` to ensure that the items are only separated by newlines and not by spaces too:
```bash
$ ls
'a b.txt' c.txt
$ langfiles=$(ls)
$ for file in $langfiles; do echo ">$file<"; done
>a<
>b.txt<
>c.txt<
```
But as there is currently no case for both `for` loops in this script where this might happen this is just a suggestion.

Brest regards
@lscharmer

pbechtold committed:
Fix duplicate identifier detection script

Simplify syntax and improve message clarity in the duplicate identifier detection script. Adjust exit statuses for better script behavior.

pbechtold commented:
hi @lscharmer, thank you very much for your great suggestions!
I have adopted your changes except for https://github.com/ILIAS-eLearning/ILIAS/pull/7802#discussion_r1713796356, as otherwise only exit code 0 is output here. 

Thanks a lot!

Best regards
@pbechtold

lscharmer commented on scripts/check_langfile_duplicate_identifier.sh:
```diff
@@ -0,0 +1,35 @@
+for file in $langfiles
+do
+  # Find the duplicate identifiers
+  duplicate_identifiers=$(sed -n '/<!-- language file start -->/,$p' $file | awk -F '#:#' '{print tolower($2)}' | sort -f | uniq -d)
```
Sorry, I missed this in the previous review: Please quote the variable `$file` to prevent argument expansion:

```suggestion
  duplicate_identifiers=$(sed -n '/<!-- language file start -->/,$p' "$file" | awk -F '#:#' '{print tolower($2)}' | sort -f | uniq -d)
```

If this is changed, I happily approve this PR :)

pbechtold replied:
@lscharmer Thank you!

lscharmer commented:
> hi @lscharmer, thank you very much for your great suggestions! I have adopted your changes except for [#7802 (comment)](https://github.com/ILIAS-eLearning/ILIAS/pull/7802#discussion_r1713796356), as otherwise only exit code 0 is output here.
> 
> Thanks a lot!
> 
> Best regards @pbechtold

Hi @pbechtold,
thank you for applying these change requests.
@https://github.com/ILIAS-eLearning/ILIAS/pull/7802#discussion_r1713796356 I just meant that you can replace `exit_status=1` with `exit 1`, but as this is optional, the current solution is fine for me :)

Best regards
@lscharmer

pbechtold committed:
Update scripts/check_langfile_duplicate_identifier.sh

Co-authored-by: lscharmer <52695099+lscharmer@users.noreply.github.com>

lscharmer commented:
Hi @matthiaskunkel,
this PR looks good to me. When the violations are fixed this can be merged.

Best regards
@lscharmer

matthiaskunkel commented:
Hi @lscharmer, thanks for the review and your okay.

Best, Matthias
